console-subscriber/tests/support/state.rs

```diff
@@ -0,0 +1,139 @@
+use std::fmt;
+
+use tokio::sync::broadcast::{
+    self,
+    error::{RecvError, TryRecvError},
+};
+
+/// A step in the running of the test
+#[derive(Clone, Debug, PartialEq, PartialOrd)]
+pub(super) enum TestStep {
+    /// The overall test has begun
+    Start,
+    /// The instrument server has been started
+    ServerStarted,
+    /// The client has connected to the instrument server
+    ClientConnected,
+    /// The future being driven has completed
+    TestFinished,
+    /// The client has finished recording updates
+    UpdatesRecorded,
+}
+
+impl fmt::Display for TestStep {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        (self as &dyn fmt::Debug).fmt(f)
+    }
+}
+
+/// The state of the test.
+///
+/// This struct is used by various parts of the test framework to wait until
+/// a specific test step has been reached and advance the test state to a new
+/// step.
+pub(super) struct TestState {
+    receiver: broadcast::Receiver<TestStep>,
+    sender: broadcast::Sender<TestStep>,
+    step: TestStep,
+}
+
+impl TestState {
+    pub(super) fn new() -> Self {
+        let (sender, receiver) = broadcast::channel(1);
+        Self {
+            receiver,
+            sender,
+            step: TestStep::Start,
+        }
+    }
+
+    /// Block asynchronously until the desired step has been reached.
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if the underlying channel gets closed.
+    pub(super) async fn wait_for_step(&mut self, desired_step: TestStep) {
+        loop {
+            if self.step >= desired_step {
+                break;
+            }
+
+            match self.receiver.recv().await {
+                Ok(step) => self.step = step,
+                Err(RecvError::Lagged(_)) => {
+                    // we don't mind being lagged, we'll just get the latest state
+                }
+                Err(RecvError::Closed) => {
+                    panic!("failed to receive current step, waiting for step: {desired_step}, did the test abort?");
+                }
+            }
+        }
+    }
+
+    /// Check whether the desired step has been reached without blocking.
+    pub(super) fn try_wait_for_step(&mut self, desired_step: TestStep) -> bool {
```

Comment:
nit: i get why this ended up being named "try_wait_for_step", because it's a version of `wait_for_step` that doesn't wait, but...couldn't we just call it
```suggestion
    /// Returns `true` if the current step is `desired_step` or later.
    pub(super) fn is_step(&mut self, desired_step: TestStep) -> bool {
```
(in particular, I would expect a function called `try_{foo}` to return a `Result`...)

Reply:
Yep, agree with this.